app/models/cart_item/validator.rb

```diff
@@ -10,19 +10,23 @@ def initialize(cart_item)
 
         def validate!
           raise CartItem::InvalidProductIdError unless product_id_valid?
-          raise CartItem::InvalidQuantityError unless quantity_valid?
+          raise CartItem::InvalidQuantityTypeError unless quantity_type_valid?
+          raise CartItem::InvalidQuantityValueError unless quantity_value_valid?
         end
 
         private
 
         def product_id_valid?
-          product = @cart_item.product
-          !product.nil?
+          !@cart_item.product.nil?
```

Comment:
nil is one of the worst part of Ruby and programming itself. 

https://robots.thoughtbot.com/rails-refactoring-example-introduce-null-object
https://en.wikipedia.org/wiki/Null_Object_pattern

and a lot of other articles in google. Nil == Evil